File: Classes/Progress Bar/DTPolishedProgressBar.m
//
//  DTPolishedProgressBar.m
//  SimpleComic
//
//  Created by Alexander Rauchfuss on 11/7/09.
//  Copyright 2009 Dancing Tortoise Software. All rights reserved.
//

#import "DTPolishedProgressBar.h"
#import "TSSTImageUtilities.h"


@interface DTPolishedProgressBar ()

@property (strong, nonatomic) NSColor *backgroundColor;
@property (strong, nonatomic) NSColor *barBackgroundColor;
@property (strong, nonatomic) NSColor *barProgressColor;
@property (strong, nonatomic) NSColor *borderColor;

@end

@implementation DTPolishedProgressBar

@synthesize progressRect, horizontalMargin, leftToRight, maxValue, currentValue, numberStyle;

- (instancetype)initWithFrame:(NSRect)frame
{
    self = [super initWithFrame:frame];
    if (self)
    {
        self.backgroundColor = [NSColor colorWithCalibratedRed: 1 green: 1 blue: 1 alpha: 0.55];
        self.barBackgroundColor = [NSColor colorWithRed:0.75 green:0.75 blue:0.75 alpha:1];
        self.barProgressColor = [NSColor colorWithDeviceRed:0.44 green:0.44 blue:0.44 alpha:1];
        self.borderColor = [NSColor colorWithRed:0 green:0 blue:0 alpha:.25];
        
		self.numberStyle = @{NSFontAttributeName: [NSFont fontWithName:@"Helvetica Neue" size:10],
							 NSForegroundColorAttributeName: [NSColor colorWithDeviceWhite: 0.2 alpha: 1]};
        
		self.horizontalMargin = 5;
        self.leftToRight = YES;
        
        [self setFrameSize: frame.size];
		[self addObserver: self forKeyPath: @"leftToRight" options: 0 context: nil];
		[self addObserver: self forKeyPath: @"currentValue" options: 0 context: nil];
		[self addObserver: self forKeyPath: @"maxValue" options: 0 context: nil];
    }
    return self;
}


- (void) dealloc
{
	[self removeObserver: self forKeyPath: @"leftToRight"];
	[self removeObserver: self forKeyPath: @"currentValue"];
	[self removeObserver: self forKeyPath: @"maxValue"];
	[self removeTrackingArea: [self trackingAreas][0]];
	
}



- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
{
	[self setNeedsDisplay: YES];
}


- (BOOL)mouseDownCanMoveWindow
{
	return NO;
}

/*
 Draws the progressbar.
 */
- (void)drawRect:(NSRect)rect
{
    NSString *totalString = [@(maxValue) stringValue];
    NSString *progressString = [@(self.currentValue + 1) stringValue];
    NSString *leftString, *rightString;
    
    NSRect bounds = [self bounds];
    NSRect indicatorRect = NSMakeRect(0, NSHeight(bounds) - 9, 2, 9);
    NSRect barRect = NSMakeRect(0, NSHeight(bounds) - 5, NSWidth(bounds), 5);
    NSRect fillRect;
    
    NSSize leftSize, rightSize;
    
    // Draw background
    [self.backgroundColor set];
    NSRectFillUsingOperation(bounds, NSCompositeSourceOver);

    // Draw bar background
    [self.barBackgroundColor set];
    fillRect = barRect;
    NSRectFill(fillRect);

    // Determine label positions and progress rect size+position
    if(leftToRight)
    {
        fillRect.size.width = NSWidth(bounds) * (currentValue + 1) / maxValue;
        indicatorRect.origin.x = round(NSWidth(fillRect)-2);
        
        leftString = progressString;
        rightString = totalString;
    }
    else
    {
		fillRect.size.width = NSWidth(bounds) * (currentValue + 1) / maxValue;
		fillRect.origin.x = round(NSWidth(bounds) - NSWidth(fillRect));
        indicatorRect.origin.x = NSMinX(fillRect);
        
        leftString = totalString;
        rightString = progressString;
    }
    
    leftSize = [leftString sizeWithAttributes: self.numberStyle];
    rightSize = [leftString sizeWithAttributes: self.numberStyle];
	
    // Draw progress
    [self.barProgressColor set];
    NSRectFill(fillRect);
    
    // Draw indicator
    [[NSColor blackColor] set];
    NSRectFill(indicatorRect);

    // Draw labels
    NSRect leftStringRect = NSMakeRect(self.horizontalMargin, NSMinY(bounds), leftSize.width, 17);
	[leftString drawInRect:leftStringRect withAttributes: self.numberStyle];
    
    NSRect rightStringRect = NSMakeRect(NSWidth(bounds) - self.horizontalMargin - rightSize.width, NSMinY(bounds), rightSize.width, 17);
    [rightString drawInRect:rightStringRect withAttributes: self.numberStyle];
    
    // Draw borders
    NSRect leftBorder = NSMakeRect(0, 0, 1, NSHeight(bounds));
    NSRect rightBorder = NSMakeRect(NSWidth(bounds)-1, 0, 1, NSHeight(bounds));
    
    [self.borderColor set];
    
    NSRectFillUsingOperation(leftBorder, NSCompositeSourceOver);
    NSRectFillUsingOperation(rightBorder, NSCompositeSourceOver);
}


/*
 This method has been over-ridden to change the progressRect porperty every time the
 progress view is re-sized.
 */
- (void)setFrameSize:(NSSize)size
{
    self.progressRect = NSMakeRect(0, 0, size.width, size.height);
    [super setFrameSize: size];
}



/*
 If there has been a mouse tracking area added to this view it will be updated
 every time the progress bar is re-sized.
 The tracking area is based on the progressRect property.
 */
- (void)updateTrackingAreas
{
	if([[self trackingAreas] count] == 0)
	{
		return;
	}
	
	NSTrackingArea * oldArea = [self trackingAreas][0];
	[self removeTrackingArea: oldArea];
	
	NSTrackingArea * newArea = [[NSTrackingArea alloc] initWithRect: progressRect 
															options: [oldArea options]
															  owner: [oldArea owner]
														   userInfo: [oldArea userInfo]];
	[self addTrackingArea: newArea];
}


/*
 Changes the currentValue based on where the user clicks.
 */
- (void)mouseDown:(NSEvent *)event
{
    NSPoint cursorPoint = [self convertPoint: [event locationInWindow] fromView: nil];
    if(NSMouseInRect( cursorPoint, progressRect, [self isFlipped]))
    {
		self.currentValue = [self indexForPoint: cursorPoint];
    }
}



/*
 Kind of surprised that the mouseDown event method would not refresh.  
 Not enough code to be worth abstracting.
 */
- (void)mouseDragged:(NSEvent *)event
{
    NSPoint cursorPoint = [self convertPoint: [event locationInWindow] fromView: nil];
    if(NSMouseInRect( cursorPoint, progressRect, [self isFlipped]))
    {
		self.currentValue = [self indexForPoint: cursorPoint];

        [[NSNotificationCenter defaultCenter] postNotificationName:@"SCMouseDragNotification" object:self];
    }
}


/*
 Translates a point within the view to an index between 0 and maxValue.
 Progress indicator direction affects the index.
 */
- (NSInteger)indexForPoint:(NSPoint)point
{
    NSInteger index;
    if(leftToRight)
    {
        index = (point.x - NSMinX(progressRect)) / NSWidth(progressRect) * maxValue;
    }
    else
    {
        index = (NSMaxX(progressRect) - point.x) / NSWidth(progressRect) * maxValue;
    }
    index = index >= maxValue ? maxValue - 1 : index;
    return index;
}

@end

File: Classes/Progress Bar/DTPolishedProgressBar.swift
/*
 Copyright (c) 2006-2009 Dancing Tortoise Software
 Created by Alexander Rauchfuss

 Permission is hereby granted, free of charge, to any person
 obtaining a copy of this software and associated documentation
 files (the "Software"), to deal in the Software without
 restriction, including without limitation the rights to use,
 copy, modify, merge, publish, distribute, sublicense, and/or
 sell copies of the Software, and to permit persons to whom the
 Software is furnished to do so, subject to the following
 conditions:

 The above copyright notice and this permission notice shall be
 included in all copies or substantial portions of the Software.

 DTPolishedProgressBar.swift
*/
//
//  Created by C.W. Betts on 10/24/15.
//

import Cocoa

private let backgroundColor = NSColor(calibratedRed: 1, green: 1, blue: 1, alpha: 0.55)
private let barBackgroundColor = NSColor(red: 0.75, green: 0.75, blue: 0.75, alpha: 1)
private let barProgressColor = NSColor(deviceRed: 0.44, green: 0.44, blue: 0.44, alpha: 1)
private let borderColor = NSColor(red:0, green: 0, blue: 0, alpha: 0.25)

/// The font attributes of the progress numbers.
private let numberStyle: [NSAttributedString.Key: Any] = [.font: NSFont.systemFont(ofSize: 10),
														 .foregroundColor: NSColor(deviceWhite: 0.2, alpha: 1)]


/**
Configurable progress bar. Allows the setting of various style attributes.
Progress direction can be set.
*/
class DTPolishedProgressBar: NSView {

	override init(frame frameRect: NSRect) {
		super.init(frame: frameRect)
		setFrameSize(frameRect.size)
	}
	
	required init?(coder: NSCoder) {
		super.init(coder: coder)
		setFrameSize(frame.size)
	}

/*
List of replacements for the highcontrast flag
Highlight: NSColor if nil then layout is slightly shifted.
barFill: This is the gradient of the empty portion of the progress bar
progressFill: This is the gradient of the filled portion of the progress bar.
shadow: This is the gradient that give the illusion of depth.
textStyle: Dictionary of string attributes.
*/

	/// The direction of the progress bar.
	@IBInspectable
	dynamic var leftToRight: Bool = true {
		didSet {
			needsDisplay = true
		}
	}
	/// The maximum value of the progress bar.
	@IBInspectable
	dynamic var maxValue: Int = 0 {
		didSet {
			needsDisplay = true
		}
	}
	/// The progress bar is filled to this level.
	@IBInspectable
	dynamic var currentValue: Int = 0 {
		didSet {
			needsDisplay = true
		}
	}
	
	/// This is the section of the view. Users can mouse over and click here.
	@objc private(set) var progressRect = NSRect()
	
	/// How much room is given for the text on either side.
	private var horizontalMargin: CGFloat = 5

	/// Translates a point within the view to an index between `0` and `maxValue`.
	/// Progress indicator direction affects the index.
	@objc(indexForPoint:) func index(for point: NSPoint) -> Int {
		var index: Int
		if leftToRight {
			index = Int((point.x - progressRect.minX) / progressRect.width * CGFloat(maxValue))
		} else {
			index = Int((progressRect.maxX - point.x) / progressRect.width * CGFloat(maxValue))
		}
		index = index >= maxValue ? maxValue - 1 : index
		// Because maxValue might be 0, making index = -1
		guard index >= 0 else {
			return 0
		}
		return index
	}

	/// Draws the progressbar.
	override func draw(_ dirtyRect: NSRect) {
		let totalString = String(maxValue)
		let progressString = String(currentValue + 1)
		let leftString: String
		let rightString: String
		
		let bounds2 = self.bounds
		var indicatorRect = NSMakeRect(0, bounds2.height - 9, 2, 9)
		let barRect = NSMakeRect(0, bounds2.height - 5, bounds2.width, 5)
		
		// Draw background
		backgroundColor.set()
		bounds2.fill(using: .sourceOver)
		
		// Draw bar background
		barBackgroundColor.set()
		var fillRect = barRect
		fillRect.fill()
		
		// Determine label positions and progress rect size+position
		if leftToRight {
			fillRect.size.width = bounds2.width * CGFloat(currentValue + 1) / CGFloat(maxValue)
			indicatorRect.origin.x = round(fillRect.width - 2)
			
			leftString = progressString
			rightString = totalString
		} else {
			fillRect.size.width = bounds2.width * CGFloat(currentValue + 1) / CGFloat(maxValue)
			fillRect.origin.x = round(bounds2.width - fillRect.width)
			indicatorRect.origin.x = fillRect.minX
			
			leftString = totalString
			rightString = progressString
		}
		
		let leftSize = leftString.size(withAttributes: numberStyle)
		let rightSize = rightString.size(withAttributes: numberStyle)
		
		// Draw progress
		barProgressColor.set()
		fillRect.fill()
		
		// Draw indicator
		NSColor.black.set()
		indicatorRect.fill()
		
		// Draw labels
		let leftStringRect = NSRect(x: horizontalMargin, y: bounds2.minY, width: leftSize.width, height: 17)
		leftString.draw(in: leftStringRect, withAttributes: numberStyle)
		
		let rightStringRect = NSRect(x: bounds2.width - horizontalMargin - rightSize.width, y: bounds2.minY, width: rightSize.width, height: 17)
		rightString.draw(in: rightStringRect, withAttributes: numberStyle)
		
		// Draw borders
		let leftBorder = NSRect(x: 0, y: 0, width: 1, height: bounds2.height)
		let rightBorder = NSRect(x: bounds2.width - 1, y: 0, width: 1, height: bounds2.height)
		
		borderColor.set()
		
		leftBorder.fill(using: .sourceOver)
		rightBorder.fill(using: .sourceOver)
	}
	
	/// This method has been over-ridden to change the progressRect porperty every time the
	/// progress view is re-sized.
	override func setFrameSize(_ size: NSSize) {
		progressRect = NSRect(origin: .zero, size: size)
		super.setFrameSize(size)
	}
	
	/// If there has been a mouse tracking area added to this view it will be updated
	/// every time the progress bar is re-sized.
	/// The tracking area is based on the `progressRect` property.
	override func updateTrackingAreas() {
		defer {
			super.updateTrackingAreas()
		}
		guard trackingAreas.count != 0 else {
			return
		}
		
		let oldArea = trackingAreas.first!
		removeTrackingArea(oldArea)
		
		let newArea = NSTrackingArea(rect: progressRect, options: oldArea.options, owner: oldArea.owner, userInfo: oldArea.userInfo)
		addTrackingArea(newArea)
	}
	
	override var mouseDownCanMoveWindow: Bool {
		return false
	}
	
	/// Changes the `currentValue` based on where the user clicks.
	override func mouseDown(with theEvent: NSEvent) {
		let cursorPoint = convert(theEvent.locationInWindow, from: nil)
		if NSMouseInRect(cursorPoint, progressRect, isFlipped) {
			self.currentValue = index(for: cursorPoint)
		}
	}
	
	/// Kind of surprised that the mouseDown event method would not refresh.
	/// Not enough code to be worth abstracting.
	override func mouseDragged(with theEvent: NSEvent) {
		let cursorPoint = convert(theEvent.locationInWindow, from: nil)
		if NSMouseInRect(cursorPoint, progressRect, isFlipped) {
			self.currentValue = index(for: cursorPoint)
			
			NotificationCenter.default.post(name: .TSSTMouseDrag, object: self)
		}
	}
	
	deinit {
		removeTrackingArea(trackingAreas[0])
	}
}
